include/dali/core/geom/vec.h

```diff
@@ -0,0 +1,378 @@
+    for (size_t i = 0; i < N; i++)
+      v[i] = scalar;
+  }
+
+  template <typename... Args, typename =
+            std::enable_if_t<dali::all_of<std::is_convertible<Args, T>::value...>::value>>
+  DALI_HOST_DEV
+  constexpr vec_base(T v0, T v1, Args... args) : v{v0, v1, T(args)... } {}
+  T v[N];
+
+  T &operator[](size_t i) { return v[i]; }
+  const T &operator[](size_t i) const { return v[i]; }
+};
+
+template <typename T>
+struct vec_base<1, T> {
+  union {
+    T v[1];
+    T x;
+  };
+
+  DALI_HOST_DEV
+  constexpr vec_base() : x() {}
+  DALI_HOST_DEV
+  constexpr vec_base(const T &x) : x(x) {}  // NOLINT
+};
+
+template <typename T>
+struct vec_base<2, T> {
+  union {
+    T v[2];
+    struct { T x, y; };
+  };
+
+  DALI_HOST_DEV
+  constexpr vec_base() : x() {}
+  DALI_HOST_DEV
+  constexpr vec_base(const T &scalar) : x(scalar), y(scalar) {}  // NOLINT
+  DALI_HOST_DEV
+  constexpr vec_base(const T &x, const T &y) : x(x), y(y) {}
+};
+
+template <typename T>
+struct vec_base<3, T> {
+  union {
+    T v[3];
+    struct { T x, y, z; };
+  };
+
+  DALI_HOST_DEV
+  constexpr vec_base() : x() {}
+  DALI_HOST_DEV
+  constexpr vec_base(const T &scalar) : x(scalar), y(scalar), z(scalar) {}  // NOLINT
+  DALI_HOST_DEV
+  constexpr vec_base(const T &x, const T &y, const T &z) : x(x), y(y), z(z) {}
+};
+
+template <typename T>
+struct vec_base<4, T> {
+  union {
+    T v[4];
+    struct { T x, y, z, w; };
+  };
+
+  DALI_HOST_DEV
+  constexpr vec_base() : x() {}
+  DALI_HOST_DEV
+  constexpr vec_base(const T &scalar) : x(scalar), y(scalar), z(scalar), w(scalar) {}  // NOLINT
+  DALI_HOST_DEV
+  constexpr vec_base(const T &x, const T &y, const T &z, const T &w) : x(x), y(y), z(z), w(w) {}
+};
+
+template <size_t N, typename T>
+struct vec : vec_base<N, T> {
+  static_assert(std::is_standard_layout<T>::value,
+                "Cannot create a vector ofa non-standard layout type");
+  using vec_base<N, T>::vec_base;
+  using vec_base<N, T>::v;
+
+  DALI_HOST_DEV
+  constexpr T &operator[](size_t i) { return v[i]; }
+  DALI_HOST_DEV
+  constexpr const T &operator[](size_t i) const { return v[i]; }
+
+  template <typename U>
+  DALI_HOST_DEV
+  constexpr vec<N, U> cast() const {
+    vec<N, U> ret;
+    for (size_t i = 0; i < N; i++) {
+      ret.v[i] = static_cast<U>(v[i]);
+    }
+    return ret;
+  }
+
+  DALI_HOST_DEV constexpr size_t size() const { return N; }
+  DALI_HOST_DEV constexpr vec operator+() const { return *this; }
+  DALI_HOST_DEV constexpr T *begin() { return &v[0]; }
+  DALI_HOST_DEV constexpr const T *cbegin() const { return &v[0]; }
+  DALI_HOST_DEV constexpr const T *begin() const { return &v[0]; }
+  DALI_HOST_DEV constexpr T *end() { return &v[N]; }
+  DALI_HOST_DEV constexpr const T *cend() const { return &v[N]; }
+  DALI_HOST_DEV constexpr const T *end() const { return &v[N]; }
+
+  DALI_HOST_DEV constexpr auto length_square() const {
+    decltype(v[0]*v[0] + v[0]*v[0]) ret = v[0]*v[0];
```

Comment:
```suggestion
    decltype(v[0]*v[0]  * N) ret = v[0]*v[0];
```

Reply:
N has nothing to do with it - it's `size_t` and it would bloat the type. Actually, the addition doesn't matter for any fundamental type, but could make a difference with some user-defined type.